Add the classical approach using linear programming

Iodine98 opened the pull request:
This PR includes the classical approach to the scheduling problem by defining the constraints and adding these to a linearly defined problem using [pulp](https://coin-or.github.io/pulp/).

Iodine98 committed:
- Add Jupyter package (for Jetbrains based IDEs) and pulp package for linear programming.
- Add draft classical approach using linear programming.

Iodine98 committed:
Add qiskit and pexpect as dependencies.

Change `value` import from pulp to `pulp_value` to avoid name shadowing.

Iodine98 committed:
Merge commit 'c135bf03370dc401be3da7d4f98dca731c7cc5fd' into 1-add-classical-approach

Iodine98 commented:
Please check carefully if the notebook still runs as expected. I did not check if the backends of the quantum computers run, because I do not have an account at Amazon, IBM, etc.

Iodine98 commented:
The linting check failed on a package that I did not introduce:

![image](https://github.com/schubergphilis/qnt-planning/assets/12817322/d921e9cf-5779-4b4e-9470-aafd0b3c3caa)
****

costastf commented:
There are two depenencies jupyter and pexpect that are added to pipenv but not used.

costastf commented:
> Please check carefully if the notebook still runs as expected. I did not check if the backends of the quantum computers run, because I do not have an account at Amazon, IBM, etc.

The backends are irrelevant of the usage of probabilities. If the code works up until the export of the bound_c then the probabilities work fine.

huisman commented on planning.ipynb:
````diff
@@ -211,15 +211,17 @@
-    "```\n"
+    "```\n",
+    "\n",
+    "Do not forget to select your interpreter (kernel) before running the other cells in the notebook!\n"
````
The step of selecting/creating the correct kernel is indeed missing from the description.
Admittedly, I didn´t use VSCode so maybe the Jupyter extension already takes care of this, but I had to create and select a kernel for the pipenv environment I used.

Iodine98 replied:
I have added that instruction in the latest commit.

huisman commented:
> The backends are irrelevant of the usage of probabilities. If the code works up until the export of the bound_c then the probabilities work fine.

This PR does not touch the probabilities as far as I can see?

costastf commented:
> > The backends are irrelevant of the usage of probabilities. If the code works up until the export of the bound_c then the probabilities work fine.
> 
> This PR does not touch the probabilities as far as I can see?

No it does not, i was just doing a hundred different things at the same time and was thinking of a comment for the previous PR 😊

costastf committed:
Fix pipeline dependencies. Validate code and add headers.

Iodine98 committed:
Merge pull request #8 from schubergphilis/fix-add-classical-approach

Fix pipeline dependencies. Validate code and add headers.

Iodine98 commented:
The fixing branch has been merged in the current one. Everything runs after merge. Can @costastf and/or @huisman verify that it works; by approving the review? If it has been approved, I will merge it.

costastf commented:
Wait, why do we have Jupiter on the main dependencies?

Iodine98 commented:
> Wait, why do we have Jupyter on the main dependencies?

Because I tried executing the notebook on a Jetbrains IDE (Pycharm) and it asked me to install Jupyter to run the notebook. Therefore, I think Jupyter should be added as a (dev-)(dependency.

Iodine98 committed:
Set Jupyter dependency as dev-dependency instead.

costastf commented:
I would think that that should be installed as an addon on pycharm itself and not be brought in as a dependency of the project.

costastf commented:
The exact same approach that has been taken for vscode. The Jupiter addon is installed on the editor and not part of the project.

Iodine98 commented:
> The exact same approach that has been taken for vscode. The Jupiter addon is installed on the editor and not part of the project.

Unfortunately, that is not how Jetbrains implemented Jupyter notebook support in Pycharm. They have support [built-in](https://www.jetbrains.com/help/pycharm/jupyter-notebook-support.html#ui) given that the `Jupyter` package has been installed in the current Python environment.

costastf commented:
> > The exact same approach that has been taken for vscode. The Jupiter addon is installed on the editor and not part of the project.
> 
> Unfortunately, that is not how Jetbrains implemented Jupyter notebook support in Pycharm. They have support [built-in](https://www.jetbrains.com/help/pycharm/jupyter-notebook-support.html#ui) given that the `Jupyter` package has been installed in the current Python environment.

On  my Pycharm i do not have jupyter installed as a dependency of the project while having this addon installed on the editor 
![Screenshot 2023-12-18 at 14 38 54](https://github.com/schubergphilis/qnt-planning/assets/4638059/fa37fdf0-6df1-4f7f-af74-30caf56eb673)
and everything works as expected. Could you please validate that this addon in enabled on your IDE?

Iodine98 committed:
- Removed Jupyter dependency from Pipfile
- Changed one cell from os.system to markdown to maintain consistency between installing dependencies in Pipenv cells. Also to avoid accidental installation of packages.

Iodine98 committed:
Removed linting error and relocked pipfile to be sure.

Iodine98 committed:
Specified versions in Pipfile for packages and dev-packages, and secondd attempt to relock Pipfile

Iodine98 committed:
Specified NBQA version to ensure compatability.

Iodine98 commented:
@costastf, there's something with the `pexpect` dependency that I cannot seem to solve. On my local machine, it does not ask for this module. Can you have a look at my latest commit?

costastf commented:
> @costastf, there's something with the `pexpect` dependency that I cannot seem to solve. On my local machine, it does not ask for this module. Can you have a look at my latest commit?

on it

costastf committed:
Fix dependencies. Add configuration files for quality tools.
